### PR TITLE
Instancing

### PR DESCRIPTION
**Pull request by guycalledfrank**
Draws multiple objects with a single draw call (ANGLE_instanced_arrays). Usually only beneficial when drawing hundreds/thousands of low-poly objects. It is currently mostly focused on drawing same objects with different matrix, but per-instance parameters can be made customizable in future. There are 3 ways of using instancing in this PR:

1. Explicit static - allocate instances once and draw.
-- meshInstance.instancingData = new pc.InstancingData(numInstances);
-- Fill meshInstance.instancingData.buffer with 4x4 matrices (numInstances * 16 floats);

2. Explicit dynamic - update matrices at runtime when needed.
-- meshInstance.instancingData = new pc.InstancingData(numInstances, true);
-- Fill meshInstance.instancingData.buffer at any time;
-- Call meshInstance.instancingData.update() after changing the buffer;

3. Automatic - multiple instances of the same object will be automatically drawn via instancing.
-- toggle it with device.enableAutoInstancing (false by default);
-- device.autoInstancingMaxObjects is the maximum allowed object count for auto instancing (default is 16384);

All instancing modes require material.useInstancing to be set.

Automatic can be a bit slower than explicit dynamic and even slower than wisely used explicit static.

Instancing is not a silver bullet - it really depends on the balance between draw call count and the complexity of each mesh.

Additional changes:
- VertexBuffer constructor now accepts initialData value;
- VertexBuffer has setData method as an alternative to lock/unlock (if you already have the buffer, you don't have to lock -> copy -> unlock, you can just set it).

test scene:
http://playcanvas.com/designer/339478/scene/356093

10k

### CODE REVIEW

File: src/graphics/graphics_device.js [hunk 1]
```diff
@@ -101,8 +101,11 @@ pc.extend(pc, function () {
         this.indexBuffer = null;
         this.vertexBuffers = [];
         this.precision = "highp";
+        this.enableAutoInstancing = true;
```
Comment (willeastcott) [inline src/graphics/graphics_device.js]: In your PR description, you say this is false by default. Which did you intend?

**Comment by guycalledfrank** [comment]
Fixed

File: src/graphics/graphics_device.js [hunk 2]
```diff
@@ -912,10 +934,20 @@ pc.extend(pc, function () {
...
+                    this.attributesInvalidated = true;
+                } else {
+                    gl.drawElements(this.glPrimitive[primitive.type],
+                                    primitive.count,
+                                    this.indexBuffer.glFormat,
+                                    primitive.base * 2);
+                }
             } else {
                 gl.drawArrays(this.glPrimitive[primitive.type],
                               primitive.base,
```
Comment (willeastcott) [inline src/graphics/graphics_device.js]: So it's not possible to render unindexed primitive instances?

Comment (guycalledfrank) [inline src/graphics/graphics_device.js]: When do we use unindexed primitives in engine?

Comment (willeastcott) [inline src/graphics/graphics_device.js]: The FBX -> JSON tools only generates indexed, but you can create pc.Mesh objects that are unindexed procedurally.

**Comment by guycalledfrank** [comment]
Now works for non-indexed objects too

**Pull request merged**